File: clojure/ldc/test/car_tax.clj
(ns ldc.test.car_tax
	(:use clojure.contrib.test-is)
	(:use ldc.car_tax))

(defstruct car :owner :speed :type :license-plate)
(defstruct taxed_car :owner :speed :type :license-plate :tax)

(deftest test-0-car
 (is (= [] (car-tax []))))

(deftest test-1-car-speed-100
 (def my_car (struct car "Christophe" 100 "Ferrari" "0123456"))
 (def my_taxed_car (car-tax [my_car]))
 (def expected_car [(struct taxed_car "Christophe" 100 "Ferrari" "0123456" 10 )])
 (is (= expected_car my_taxed_car)))


(deftest test-2-cars-speed-110
 (def my_car (struct car "Christophe" 110 "Ferrari" "0123456"))
 (def my_taxed_car (car-tax [my_car my_car]))
 (def expected_car (struct taxed_car "Christophe" 110 "Ferrari" "0123456" 210 ))
 (is (= [expected_car expected_car] my_taxed_car)))

(deftest test-2-cars-speed-10-110
 (def my_car_10 (struct car "Christophe" 10 "Ferrari" "0123456"))
 (def my_car_110 (struct car "Christophe" 110 "Ferrari" "0123456"))
 (def my_taxed_car (car-tax [my_car_10 my_car_110]))
 (def expected_car_10 (struct taxed_car "Christophe" 10 "Ferrari" "0123456" 0 ))
 (def expected_car_110 (struct taxed_car "Christophe" 110 "Ferrari" "0123456" 210 ))
 (is (= [expected_car_10 expected_car_110] my_taxed_car)))

File: clojure/car-tax/src/main/java/ldc/cartax/CarTax.java
package ldc.cartax;

import java.io.IOException;
import java.util.Arrays;

import clojure.lang.IPersistentMap;
import clojure.lang.ISeq;
import clojure.lang.LazySeq;
import clojure.lang.RT;
import clojure.lang.Var;

public class CarTax {

	public static void main(String[] args) throws IOException {
		RT.loadResourceScript("ldc/cartax/car_tax.clj");
		Var carTax = RT.var("ldc.cartax.car-tax", "car-tax");

		IPersistentMap carMap1 = RT.map(
				RT.keyword(null, "owner"), "Christophe", 
				RT.keyword(null, "speed"), 100, 
				RT.keyword(null, "type"), "Ferrari", 
				RT.keyword(null, "licence-plate"), "123456");
		IPersistentMap carMap2 = RT.map(
				RT.keyword(null, "owner"), "Bob",
				RT.keyword(null, "speed"), 250, 
				RT.keyword(null, "type"), "Porsche", 
				RT.keyword(null, "licence-plate"), "1234567");
		ISeq cars = RT.seq(Arrays.asList(carMap1, carMap2));
		LazySeq results = (LazySeq) carTax.invoke(cars);
		for(Object o : results){
			System.out.println(o);
		}
	}
}
